### ISSUE TITLE
OSD not working with server-side rendering frameworks like Remix 

### ISSUE THREAD

**Issue by jrydberg** [openseadragon/openseadragon#2105]
Importing the package has a lot of module side effects, trying to access `document`, `window` or `navigator`. This does not working with server-side rendering frameworks like Remix.

**Comment by iangilman** [comment]
Good point! OpenSeadragon was started long before such frameworks existed. What would be involved in refactoring it to work in that environment? Any thoughts you have would be helpful for whoever might choose to pick up this feature and implement it.

**Comment by peterjcaulfield** [comment]
I don't think that the library would need to be refactored but users would need to create a wrapper around the lib to make the code isomorphic. 

There is nice example of a React hook based integration here: https://github.com/stephenwf/use-open-seadragon

It doesn't appear to be actively maintained (still points to last OSD major version) but it's illustrative of the approach needed to make the lib work in an SSR context.

**Comment by iangilman** [comment]
@peterjcaulfield Great! Thank you for sharing this. So maybe it's a matter of updating our documentation somehow to point to this technique? Maybe we need an example page for it?

**Comment by peterjcaulfield** [comment]
Just to elaborate on this as I kind of mis-spoke a little, there are two cases to handle to ensure the lib works in an SSR context with frameworks like Remix/Next.js:

- The initial import of the lib (if the module has side effects that rely on browser globals)
- The initial render 

In the case of OSD, there are unfortunately some side effects that occur during the import of the module as the module uses a [SEAF](https://github.com/openseadragon/openseadragon/blob/master/src/openseadragon.js#L784) that contains code that [accesses globals](https://github.com/openseadragon/openseadragon/blob/master/src/openseadragon.js#L939) that only exist in a browser context.

The "easy" fix for users without any changes to OSD is to use some form of dynamic import. In Next.js you can see examples for:

- [modules](https://nextjs.org/docs/advanced-features/dynamic-import)
- [components](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr)

Both options allow the corresponding code to opt out of server-side rendering. You can either dynamically import OSD or you can dynamically import the component that uses OSD.

The other option would be to refactor the OSD lib not to have side effects that depend on browser globals during import.

**Comment by iangilman** [comment]
Thank you for explaining! I would say there's no point in having OSD render on the server at all, so the `ClientOnly` technique sounds like the right way to go. That said, if there is a reason to have it render on the server, I'd love to know what that is, and what "rendering" would result in. 

As for the documentation, I'm wondering where that should go. Perhaps another section on this page? http://openseadragon.github.io/docs/

**Comment by kaicataldo** [comment]
I just ran into this issue this past week as well. For my use case, we wrap OSD in a React component as part of a Next.js app, which server-side renders the app before hydrating it on the client. Would there be any appetite for a PR where we add guards for only the browser globals being accessed at import time? The runtime accesses would remain untouched. If so, I'd be happy to take a crack at it!

**Comment by Aiosa** [comment]
Please be careful about 'server side rendering' here, since it might mean that the server pre-renders the image data before delivering individual cases. Especially here it is a common (and older) technique, which is totally fine for OSD. Since I am not familiar with remix, I thought this issue discusses tile-prerendering...

**Comment by iangilman** [comment]
@kaicataldo Sounds like you're just talking about adding guards so OSD doesn't blow up doing the server side rendering process, right? You wouldn't actually have OSD do anything on the server side?

How invasive do you expect this would be? I'm certainly up for seeing what a PR would look like.

**Comment by kaicataldo** [comment]
Yes, exactly! Let me take a look and put up a PR so that we know exactly what this would entail as you decide if this is the direction you want to go.

**Comment by iangilman** [comment]
Sounds great! Thank you.

**Comment by kaicataldo** [comment]
Okay, finally had some time to look at this. The current script wraps each module in an IIFE to compose OpenSeadragon, and it’s during the course of this bootstrapping that browser APIs are referenced immediately at runtime. I was originally thinking we could add guards around those specific points of reference, but given that the intended goal is not to get this working on the server, but merely to be compatible with server-side runtimes, I think the simplest solution is to wrap the entire bootstrapping portion of the script in a function that gets called conditionally if `typeof window !== 'undefined'`. 

However, it looks like the current tooling is doing a simple concatenation of the files, and so the wrapper would have to be split between the first and last file in the list of concatenated files, which doesn’t feel great.

Is there any appetite to convert the source to ESM and using a bundler? We could then do a similar thing to the modern jQuery codebase and import `OpenSeadragon` into each module and compose it that way. 

If not, we could probably do something along the lines of turning the top level IIFEs into regular function declarations and then, at the end of the concatenated script, have one IIFE that runs all the previously declared functions. This function could check if it’s being executed in the browser and, if not, return early before any other setup functions are called.

I’m happy to do the legwork for this, if either of these solutions feels like the right way forward. Thanks for your work on this!

**Comment by iangilman** [comment]
Thank you for investigating this! OSD has been around for a long time, and the JS world has evolved a lot since it got started. I think it would be great to catch up with more modern ways of doing things if possible.

I'm all for converting the source to ESM and using a bundler. Ideally we should be able to allow the devs who use OSD to use it via `import` or by the old school adding to the HTML via a `script` tag.

Another thing to keep in mind is our ecosystem of plugins. Whatever we do should support the existing plugins, which are all generally designed to be loaded via a `script` tag. 

Stretch goal: it would be awesome if we could also start to support a way to have plugins that can work via the `import` mechanism (perhaps being applied to the OSD afterwards). That's probably outside the scope of this issue, but I figured I'd mention it just so you're aware.
